### ISSUE TITLE
ICC Support

### ISSUE THREAD

**Issue by SchrodingerZhu** [microsoft/snmalloc#323]
I may work on implement ICC/ICPC (intel C compiler) support for this library later.

Previously, I found that `-mcx16` is not supported by ICC. But recently, I found that with the following code
```c++
#include <cstdint>
#include <atomic>
struct alignas(2 * sizeof(std::size_t)) Linked
{
    std::size_t test[2];
};

Linked OLD = { 0, 0 };
Linked NEW = { 1, 1 };

int test(std::atomic<Linked> & a) {
    a.compare_exchange_weak(
          OLD, NEW, std::memory_order_relaxed, std::memory_order_relaxed);
}
```
ICC can generate
```c++
test(std::atomic<Linked>&):
        push      rbx                                           #12.35
        mov       esi, offset flat: OLD                         #296.9
        mov       rbx, QWORD PTR NEW[rip]                       #14.16
        mov       rcx, QWORD PTR 8+NEW[rip]                     #14.16
        mov       rax, QWORD PTR [rsi]                          #296.9
        mov       rdx, QWORD PTR [8+rsi]                        #296.9
        lock      
        cmpxchg16b QWORD PTR [rdi]                              #296.9
        mov       QWORD PTR [rsi], rax                          #296.9
        mov       QWORD PTR [8+rsi], rdx                        #296.9
        sete      al                                            #296.9
        movzx     eax, al                                       #296.9
        pop       rbx                                           #15.1
        ret                                                     #15.1
OLD:
NEW:
        .long   0x00000001,0x00000000
        .long   0x00000001,0x00000000
```
So it should be fine to support ICC now.

**Comment by SchrodingerZhu** [comment]
https://godbolt.org/z/r4sMsx6nv

**Comment by davidchisnall** [comment]
@SchrodingerZhu, do you want to close this, since you closed #326?
